flare/ase/otf.py

```diff
@@ -83,71 +82,96 @@ class ASE_OTF(OTF):
             MD = Langevin
         else:
-            raise NotImplementedError(md_engine+' is not implemented in ASE')
+            raise NotImplementedError(md_engine + " is not implemented in ASE")
 
-        self.md = MD(atoms=atoms, timestep=timestep, trajectory=trajectory,
-                     **md_kwargs)
+        self.md = MD(
+            atoms=self.atoms, timestep=timestep, trajectory=trajectory, **md_kwargs
+        )
 
-        self.atoms = atoms
         force_source = dft_source
         self.flare_calc = self.atoms.calc
 
         # Convert ASE timestep to ps for the output file.
         flare_dt = timestep / (units.fs * 1e3)
 
         super().__init__(
-            dt=flare_dt, number_of_steps=number_of_steps,
-            gp=self.flare_calc.gp_model, force_source=force_source,
-            dft_loc=dft_calc, dft_input=self.atoms, **otf_kwargs)
+            dt=flare_dt,
+            number_of_steps=number_of_steps,
+            gp=self.flare_calc.gp_model,
+            force_source=force_source,
+            dft_loc=dft_calc,
+            dft_input=self.atoms,
+            **otf_kwargs
+        )
+
+    def get_structure_from_input(self, prev_pos_init):
+        self.structure = self.atoms
+        if prev_pos_init is None:
+            self.atoms.prev_positions = np.copy(self.atoms.positions)
+        else:
+            assert len(self.atoms.positions) == len(
+                self.atoms.prev_positions
+            ), "Previous positions and positions are not same length"
+            self.atoms.prev_positions = prev_pos_init
 
     def initialize_train(self):
         super().initialize_train()
 
-        if self.md_engine == 'NPT':
+        if self.md_engine == "NPT":
             if not self.md.initialized:
                 self.md.initialize()
             else:
                 if self.md.have_the_atoms_been_changed():
                     raise NotImplementedError(
-                        "You have modified the atoms since the last timestep.")
+                        "You have modified the atoms since the last timestep."
+                    )
 
     def compute_properties(self):
-        '''
+        """
         Compute energies, forces, stresses, and their uncertainties with
             the FLARE ASE calcuator, and write the results to the
             OTF structure object.
-        '''
+        """
 
         # Reset FLARE calculator if necessary.
         if not isinstance(self.atoms.calc, FLARE_Calculator):
             self.atoms.set_calculator(self.flare_calc)
 
-        self.atoms.calc.calculate(self.atoms, self.structure)
+        self.atoms.calc.calculate(self.atoms)
 
     def md_step(self):
-        '''
+        """
         Get new position in molecular dynamics based on the forces predicted by
         FLARE_Calculator or DFT calculator
-        '''
+        """
         # Update previous positions.
         self.structure.prev_positions = np.copy(self.structure.positions)
```

Comment:
I wonder if there's a way to avoid having to keep around the structure object inside ASE OTF? Maybe after calling the inherited OTF constructor, the structure attribute can be set equal to the atoms object?

Reply:
Right, I actually set the self.structure as the atoms in ase/otf.py
```
135     def get_structure_from_input(self, prev_pos_init):
136         self.structure = self.atoms
```
Yeah this is kind of hard to notice though... The `get_structure_from_input` will be called at the init function, so no `Structure` will appear through the ase otf run